### PR TITLE
MissingPluginException on release build #452

### PR DESCRIPTION
**Pull request by Boehrsi**
Initially I tried to `-keep` just the whole package and played around with the rules. After some tries I got it working and tried to move the rules from my actual app project to the library directly. During this process I realized not the `-keep` rule for the package alone was the fix, but also the usage of `getDefaultProguardFile('proguard-android-optimize.txt')` which I applied during testing.
I also tested `getDefaultProguardFile('proguard-android.txt')` and it wasn't working, so I just diffed both and got the needed rule by trial and error. Lastly I also removed the `-keep` rule itself and funny enough it was still working.

**How to easily test in an app**

Update pubspec.yaml
``` 
# sqflite: ^1.3.2+1 replace the actual production import with the following direct Git branch import
  sqflite:
    git:
      url: git://github.com/Boehrsi/sqflite.git
      path: sqflite
      ref: 452_MissingPluginExceptionOnReleaseBuilds
``` 

Create a minified build, e.g. by just enabling it in debug builds:
```
buildTypes {
        debug {
            minifyEnabled true
            shrinkResources true
        }
    }
```

To ensure the new stuff is really used some rounds of `flutter clean` or `gradlew clean` could be helpful.

**Additional information**

- Link to the default Android ProGuard file: https://android.googlesource.com/platform/sdk/+/master/files/proguard-android.txt
- Link to the default Android optimized ProGuard file: https://android.googlesource.com/platform/sdk/+/master/files/proguard-android-optimize.txt
- ProGuard documentation: https://www.guardsquare.com/en/products/proguard/manual/usage

### CODE REVIEW

**Comment by alextekartik** [comment]
Sorry I somehow did not notice this pull request. Would it possible to apply the minify configuration on the example apps in release mode in the same project (either the one in sqflite/example or the one in sqlite_test_app or both!) so that I have a way to validate easily that it works (even if this might sound obvious for you)? Thanks!

**Comment by Boehrsi** [comment]
No problem and yep should be possible, I will have a look later today. I will update the PR 👍 .

**Comment by alextekartik** [comment]
thanks @Boehrsi !

**Comment by Boehrsi** [comment]
It seems to be more complicated to reproduce this issue than I initially thought. I tried for the last hours to provoke the error in the example app of the project and within a newly created Flutter example app. And I wasn't able to do so, so my assumption that this is a general problem is refuted. I would say this is good in general, but quite strange in regards to the issue itself, as I wasn't able to pin point the actual configuration which leads to the error.

As my own project is affected I tried to copy Gradle configurations, dependencies, plugin embedding setups and multiple other parts to find the problem, but nothing helped. If time allows I will continue testing during the weekend.

**Comment by Boehrsi** [comment]
@alextekartik I think I made some progress. It's lets say "complicated". 

Some findings:

- It's not sqflite alone, which leads to the problem
- It's related to the way Proguard / R8 are obfuscating code and renaming methods
- It seems like if another library uses or somehow references sqflite the problem occur. Some examples:
  - https://github.com/Baseflow/flutter_cached_network_image/issues/497 (I have no errors if I add this dependency)
  - https://pub.dev/packages/image_picker is the trigger for my personal project, even though I haven't found direct sqflite references within this project

This explains why only some people encounter the problem. To minimize external sources I created a fresh new test app, where I just added `sqflite: ^1.3.2+1` and `image_picker: ^0.6.7+14` and I called `openDatabase` which lead to the error. I will add this flow tomorrow to the PR, so you can have a look.

File: sqflite/android/build.gradle [hunk 1]
```diff
@@ -35,6 +35,7 @@ android {
     defaultConfig {
         minSdkVersion 16
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        consumerProguardFiles 'proguard-rules.txt'
```
Comment (Boehrsi) [inline sqflite/android/build.gradle:38]: I you comment out this line you will see the problematic `MissingPluginException(No implementation found for method getDatabasesPath on channel com.tekartik.sqflite)` exception. With the Proguard rules everything should work.

**Comment by Boehrsi** [comment]
Even though this seems to fix the symptoms somebody with more insights should probably verify why is this there a problem at all and what is the cause. I can't really tell to be honest why this rules fixes the problem.

**Comment by alextekartik** [comment]
I'm still puzzled by what to do here yet. Maybe adding documentation regarding minification (although I don't have much clue) could help. I'm going to close this pull request for now, sorry.

**Pull request closed without merging**